### ISSUE TITLE
Perguntas para a Mariana

### ISSUE THREAD

**Issue by leonardoazzi** [leonardoazzi/ipp-student-dropout#28]
Boa noite Profa., tudo bem?

Tivemos algumas dúvidas sobre o feedback do Trabalho 1 que gostaríamos de solucionar com tua ajuda. Vou trazer cada feedback e a respectiva dúvida abaixo:

1. "Uma sugestão: o fato do atributo não estar tão correlacionado com o alvo não quer dizer que não possa auxiliar na tarefa. Ele pode ajudar a distinguir as classes em complemento a outros atributos. Sugiro remover alguns atributos muito correlacionados entre si, como vimos que existe na Figura 9, mas manter os demais para uma primeira análise. Usar correlação aqui tem como ponto negativo um "data leakage", visto que está olhando para todos os dados, incluindo os que serão usados posteriormente como dados de teste. Além disso, é difícil estabelecer a priori o que seria um bom threshold de correlação. Uma forma de usar essa informação na seleção de atributos seria usar um método de filtro, em que os atributos são dispostos em um ranking e modelos são criados/avaliados com os top N atributos, sendo N um inteiro que vai incrementando de 1 ao número total de atributos."

-> Bastante pertinentes essas observações. De fato, o valor de threshold foi um grande 'chute'. Ficamos com um pouco de dúvida, porém, em qual métrica utilizar. Dado que atributos que não estão correlacionados com o alvo podem ajudar na tarefa, devemos evitar o uso da Correlação de Pearson como métrica do método de filtro? Olhamos o método SelectKBest do scikit-learn e ele tem como função que calcula o score a f_classif (https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html#sklearn.feature_selection.f_classif). Poderia ser uma métrica para score viável, ou precisamos ter outros critérios para selecionar a melhor métrica?

-> Sobre o data leakage, pensamos em adicionar a etapa de seleção de features na Pipeline, utilizando os dados de treinamento. Faz sentido?

2. "Muito bons os gráficos! [Sobre as Figuras 12-15] Sugiro tentarem colocar tudo em uma figura, com painel (a), (b), (c), (d), para facilitar comparação de modelos entre diferentes métricas."

-> Sobre esta figura com painéis, seria literalmente agrupar estas figuras existentes em um grid de sub-figuras? Como estas do link:
https://miro.medium.com/v2/resize:fit:720/format:webp/1*eCqnH2DFkcd83HiDHu9FHQ.png
Não lembro se já vi em livro texto ou na literatura, mas seria interessante criarmos uma única figura, onde o eixo x representa cada métrica? Enviamos um exemplo em anexo (metrics.png). Não tenho certeza se esta visualização é a mais interessante, pois o alinhamento de cada conjunto com o eixo x pode ser um pouco confuso. Mas se não for nada disso, poderias nos enviar um exemplo?

**Comment by leonardoazzi** [comment]
Oi, pessoal, tudo bem?

Desculpem não ter conseguido responder antes.

 

Respondo abaixo as dúvidas:
>
>  
>
>
> -> Bastante pertinentes essas observações. De fato, o valor de threshold foi 
> um grande 'chute'. Ficamos com um pouco de dúvida, porém, em qual métrica 
> utilizar. Dado que atributos que não estão correlacionados com o alvo podem 
> ajudar na tarefa, devemos evitar o uso da Correlação de Pearson como métrica 
> do método de filtro? Olhamos o método SelectKBest do scikit-learn e ele tem 
> como função que calcula o score a f_classif 
> (https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html#sklearn.feature_selection.f_classif). 
> Poderia ser uma métrica para score viável, ou precisamos ter outros critérios 
> para selecionar a melhor métrica?
>  
A análise de correlação é bastante utilizada para remover atributos redundantes, ou seja, fazer correlação dos atributos preditivos par a par, sem olhar necessariamente a relação deles com o atributo alvo. Isso seria um passo da limpeza de dados (normalmente escolhe-se um do par fortemente correlacionado para remover. O segundo passo seria usar algo na linha de um método de filtro para tirar atributos que não parecem muito relevantes para a tarefa. Realmente fazer essa limpeza com base em um threshold pré-definido é um pouco desafiador, pois como mencionei,  não só é difícil estabelecer esse threshold a priori, como também atributos com um valor de threshold um pouco mais baixo podem ainda ter alguma utilidade no treinamento do modelo. O SelectKBest é uma boa forma de encapsular ess seleção de atributos, ainda assim, ele normalmente é utilizado dentro de um loop de forma a determinar o melhor valor de k  (testa-se vários, e verifica-se o que maxima o desempenho). Nesse sentido, acredito que um método útil para vocês é o RFECV, que também faz seleção de atributos e consegue automaticamente escolher o melhor subconjunto a partir de um CV que faz treino e validação para diferentes tamanhos de subconjuntos. A vantagem é que o loop de avaliação para encontrar melhor subconjunto já está embutido na chamada do método. A desvantagem é que ele é um método que usa uma seleção de atributos embutida: A supervised learning estimator with a fit method that provides information about feature importance either through a coef_ attribute or through a feature_importances_ attribute.


Então acho que podem verificar qual se encaixa melhor no trabalho de vcs. De uma forma ou de outra, a ideia de vocês de colocar a seleção de atributos dentro do Pipeline é ótima para diminuir as chances de data leakage nesse processo.

Sobre a função a SelectKBest e a função de avaliação:  If your data consists of only categorical variables, “chi2” would be a good choice. If your data consists of both categorical and numerical variables, “mutual_info_classif” might be a better choice. If you have numerical variables and you’re looking for a linear relationship between features and target variable, “f_classif” would be a good choice.

(Fonte: https://medium.com/@Kavya2099/optimizing-performance-selectkbest-for-efficient-feature-selection-in-machine-learning-3b635905ed48)

 

 

> -> Sobre esta figura com painéis, seria literalmente agrupar estas figuras 
> existentes em um grid de sub-figuras? Como estas do link:
> https://miro.medium.com/v2/resize:fit:720/format:webp/1*eCqnH2DFkcd83HiDHu9FHQ.png
> Não lembro se já vi em livro texto ou na literatura, mas seria interessante 
> criarmos uma única figura, onde o eixo x representa cada métrica? Enviamos um 
> exemplo em anexo (metrics.png). Não tenho certeza se esta visualização é a 
> mais interessante, pois o alinhamento de cada conjunto com o eixo x pode ser 
> um pouco confuso. Mas se não for nada disso, poderias nos enviar um exemplo? 
> :P
 

Isso mesmo, a sugestão seria fazer algo assim, pois acho que fica mais fácil analisar os gráficos de forma comparativa. Mas se for o caso de juntar eles em subfiguras como no exemplo, eu sugiro manter a mesma escala no eixo y. (por exemplo, 0.5 - 1.0)

Podem decidir se preferem unir ou manter como está, não é algo que vai impactar a nota de vcs.

 

 

Sobre a formação do grupo (exposta no outro e-mail): de acordo! Aguardarei definição de vcs após entrega do T2.

 

Att.,

Mariana

**Comment by leonardoazzi** [comment]
=== NOVAS PERGUNTAS

1. Avaliação outer-loop no Nested CV deve ser igual à avaliação feita no T1? Ou seja: repeated K-Fold, variando random_seeds? Ou não precisa ser a mesma? Por ex., K-Fold com um número de folds diferente do T1 e random_seed fixo.

**Comment by leonardoazzi** [comment]
Boa tarde Prof. Mariana, tudo bem?

Temos mais uma dúvida, agora sobre o T2.

Utilizamos o Optuna com Nested Cross-Validation para a etapa de otimização de hiperparâmetros, com três algoritmos que selecionamos no spot-checking. Porém, estamos com dificuldade de entender e ter segurança na metodologia correta para a seleção de modelos.

Através de algumas leituras, entendemos que talvez selecionar um conjunto de hiperparâmetros a partir da estimativa de desempenho do Nested CV (outer loop) possa ser inconsistente, pois seu objetivo seria selecionar a melhor pipeline de modelagem - estimando o desempenho de generalização de uma forma a minimizar o viés de seleção [1]. Portanto, o objetivo desta etapa ainda não seria selecionar os melhores hiperparâmetros, mas sim oferecer uma avaliação de desempenho onde podemos visualizar qual o algoritmo com a melhor pipeline de modelagem (transformação de dados + hiperparâmetros). No nosso caso, pra ilustrar, tem sido o AdaBoost.

Vimos que o que tem se sugerido é que, após o Nested CV, deve-se realizar um novo estudo de hiperparâmetros para apenas uma pipeline de modelagem, mas utilizando todos os dados, para assim selecionar os melhores hiperparâmetros (não influenciando a estimativa de avaliação) e, por consequência, obtendo o melhor modelo para ser utilizado em novos dados.

Esta seria a abordagem mais correta? Ou existem problemas nesta metodologia?

Inicialmente, estávamos selecionando os hiperparâmetros dos melhores estudos realizados nos inner folds, mas acharmos esta abordagem um pouco inconsistente, pois é bem dificultoso selecionar um melhor estudo dentre várias rodadas, pois ela pode ter tido um bom F1-score por ter sido testada em um fold 'mais fácil'. Uma abordagem que vimos seria selecionar a média ou a moda dos hiperparâmetros, mas com o porém de que isto não faria parte do processo padrão de Nested CV.

"Among your suggestions, I think selecting hyperparameter by the best score across all folds is particularly dangerous. If one fold just happens to be much easier to predict, then the other folds won't matter for the selection, and then whichever hyperparameter performs better on this easy-to-predict fold wins, and it seems reasonably likely that hyperparameters' performances will depend on how easy the fold was to predict. Selecting instead by frequency is perhaps more robust, but it's at least not the standard approach with nested CV." [1]

Foi um pouco obscuro encontrar material conclusivo e didático o suficiente sobre este tema. Encontramos muitos fóruns com essa discussão, onde as pessoas ainda têm dificuldade de elaborar sobre a questão, pois de fato existem múltiplas formas válidas, para diferentes volumes de dados e níveis aceitáveis de 'viés'. Vou colocar aqui um resumo de um dos artigos de blog que encontramos:

"As follows:

1. An algorithm is selected based on its performance on the outer loop of nested cross-validation.
2. Then the inner-procedure is applied to the entire dataset.
3. The hyperparameters found during this final search are then used to configure a final model.
4. The final model is fit on the entire dataset." [2]

Referências:
[1] https://datascience.stackexchange.com/questions/116311/which-is-the-final-model-from-nested-cross-validation-accuracy-or-frequency
[2] https://machinelearningmastery.com/nested-cross-validation-for-machine-learning-with-python/

Enfim, agradecemos mais uma vez pela orientação. Perdão por outro e-mail longo, mas só pelo fato de escrevê-lo já me ajudou elaborar melhor o problema melhor.

**Comment by leonardoazzi** [comment]
Boa noite Prof!

Obrigado pelo retorno. Nos ajudou bastante a avançar nestes pontos.

Solucionei algumas dúvidas hoje na aula, mas agora nos surgiu uma questão, com relação à seleção de atributos que foi falada aqui (pode ser só um grande overthinking também).

    SelectKBest é uma boa forma de encapsular ess seleção de atributos, ainda
    assim, ele normalmente é utilizado dentro de um loop de forma a determinar o
    melhor valor de k (testa-se vários, e verifica-se o que maxima o desempenho).
    Nesse sentido, acredito que um método útil para vocês é o RFECV, que
    também faz seleção de atributos e consegue automaticamente escolher o
    melhor subconjunto a partir de um CV que faz treino e validação para
    diferentes tamanhos de subconjuntos. A vantagem é que o loop de avaliação
    para encontrar melhor subconjunto já está embutido na chamada do método. A
    desvantagem é que ele é um método que usa uma seleção de atributos
    embutida:_ A supervised learning estimator with a fit method that provides
    information about feature importance either through a coef_ attribute or
    through a feature_importances_ attribute._

    Então acho que podem verificar qual se encaixa melhor no trabalho de vcs. De
    uma forma ou de outra, a ideia de vocês de colocar a seleção de atributos
    dentro do Pipeline é ótima para diminuir as chances de data leakage nesse
    processo.


Entendemos que seria suficiente inserir o SelectKBest ou RFECV na pipeline que temos, que é utilizada na avaliação do Spot-Checking com validação cruzada que fizemos pro T1, junto com normalização e SMOTE. Porém, ficamos com uma dúvida: verificamos que para cada fold do spot-checking (não o do RFECV no caso), o método seleciona atributos diferentes para cada fold. Faz sentido esta seleção utilizar um subconjunto possivelmente diferente de atributos para cada fold/random_seed/modelo? Isto não poderia afetar negativamente o processo de avaliação, que buscamos comparar os modelos?

Consultamos o especialista "GePeTo", por não termos encontrado muita referência sobre os seletores de atributos neste contexto, e ele encoraja o uso, mas muito no contexto de evitar data leakage.

Se este cenário for o correto, como fazemos para selecionar os atributos 'definitivamente' após o spot-checking? Ou passamos para o T2 o mesmo dataset, sem nenhuma modificação por conta da seleção de atributos? Pergunto porque se selecionarmos 3 algoritmos, possivelmente eles selecionarão atributos diferentes.

Ficamos um pouco preocupados porque esta parte afeta muito o T2 e a interpretação dos resultados, mas espero que não seja nada de impacto grave. Aguardamos teu retorno!

Obrigado pela atenção, abraços.

(P.S.: testamos o RFECV e tivemos problemas, pois usamos o KNN no Spot Checking e ele não tem um 'importance_getter' explícito no Scikit-learn, o que nos forçaria a buscar alguma alternativa. Por ora, se possível, vamos preferir pelo SelectKBest)
